### DISCUSSION TITLE
How to generate INCON file ??

### DISCUSSION THREAD

**Discussion by teka1978** [acroucher/PyTOUGH#36] [General]
I have TOUGH2 input file and Geometry file produced on LEAPFROG, however I couldn't manage to prepare INCON file. Can you show me how to prepare initial condition?

Thanks,

**Comment by acroucher** [comment]
You can use the t2grid ``incons()`` method to create initial conditions. For example, to create simple constant initial conditions everywhere:

```python
from t2data import *
dat = t2data('mydata.dat')
values = [101.3e3, 20.]
inc = dat.grid.incons(values)
inc.write('mydata.incon')
```
If you want to create initial conditions with different values in different blocks, you can pass in a 2-D array of numbers for the 'values' parameter.

**Comment by acroucher** [comment]
Here's an example of setting up initial conditions with a simple approximate hydrostatic pressure profile (based on the distance between each block centre and the top surface of the column, which might vary from column to column). You could modify it to include e.g. a temperature gradient as well.

```python
from t2data import *

geo = mulgrid('my_geometry.dat')
dat = t2data('my_data.dat')

P0 = 101.3e3
T0 = 20.
rho = 998.
g = 9.8

vars = []
for blk in dat.grid.blocklist:
    if blk.rocktype.name == 'atmos':
        P = P0
        T = T0
    else:
        colname = geo.column_name(blk.name)
        surf = geo.column[colname].surface
        h = surf - blk.centre[2]
        P = P0 + rho * g * h
        T = T0
    vars.append([P, T])

inc = dat.grid.incons(vars)
inc..write('my_data.incon')

**Reply by teka1978** [reply]
Hi Adrian,
Is it possible to merge layers or spilit a layer?
If, so, would you mind to write a simple script ? 
Thanks.

**Reply by acroucher** [reply]
The `mulgrid` class does have a `refine_layers()` method for splitting layers. There is no method for merging layers at the moment.

**Comment by Petros2017** [comment]
Hello,  Andrian: Thank you for your contribution. Could you please assist me how import Json file(geometry file which is Json extension) to TIM or Pytough2 for visualization.

**Reply by acroucher** [reply]
Can you give me a bit more detail on what kind of JSON file it is? Is it a Waiwera JSON input file?

**Comment by acroucher** [comment]
That looks like some sort of 3D mesh file, with 3D nodes defined first, then some 3D elements? How was it generated?

If so, then you will only be able to create a geometry file if you are sure the mesh has a layer/column structure (otherwise the Mulgrid geometry file format won't support it). If that is the case, you will have to write a custom script to create the geometry file - there is nothing in PyTOUGH to convert a general 3D mesh into a Mulgrid geometry file (partly because it is not always possible). The only thing close to that is the `t2grid` `rectgeo()` method, which converts a rectangular TOUGH2 grid to a geometry.

**Comment by Petros2017** [comment]


**Reply by acroucher** [reply]
PyTOUGH doesn't have any facility for creating Voronoi grids but if you just want a rectangular grid with local refinement in the middle, you can use the `mulgrid` `refine()` method to get something very similar, but with triangular columns in the transition between coarse and fine. You can then use the 'optimize()` method to regain orthogonality in your grid (so the lines between column centres are perpendicular to the faces).

To add topography you can use the `mulgrid`  `fit_surface()` method. Changing the coordinate system to something like UTM60 is usually just a matter of shifting the grid origin, which you can do when you create the rectangular grid.

**Reply by Petros2017** [reply]
Thank you so much, for your quick response.
